Домашнее задание Модуль №15

Ulyanov-VN opened the pull request.

Ulyanov-VN committed:
Домашнее задание Модуль №15

ArturNurtdinov commented on app/src/main/java/com/example/playlistmaker/domain/interactor/GetCountryNameInteractor.kt:
```diff
@@ -0,0 +1,16 @@
+
+import java.util.Locale
+
+class GetCountryNameInteractor {
```
По идее если тут всего 1 функция и в нейминге у тебя есть глагол, то это юзкейс, а не интерактор. Но и к юзкейсам и к интеракторам нужно делать интерфейс, чтобы повысить гибкость архитектуры и тестируемость

ArturNurtdinov commented on app/src/main/java/com/example/playlistmaker/domain/interactor/PlayerInteractor.kt:
```diff
@@ -0,0 +1,97 @@
+package com.example.playlistmaker.domain.interactor
+
+import android.media.MediaPlayer
+import android.util.Log
```
В след спринте будет требование, что домейн слой не должен зависеть от классов Android SDK, так что тут лучше сразу в репозиторий вынести зависимость от Log и MediaPlayer

ArturNurtdinov commented on app/src/main/java/com/example/playlistmaker/presentation/activities/PlayerActivity.kt:
```diff
@@ -16,26 +14,47 @@ import androidx.activity.OnBackPressedCallback
     private var currentTrack: Track? = null
-    private var isPrepared = false
+
+    companion object {
```
`companion object` принято объявлять внизу класса согласно Kotlin Coding Convention (https://kotlinlang.org/docs/coding-conventions.html#class-layout)

ArturNurtdinov commented on app/src/main/java/com/example/playlistmaker/presentation/common/ThemeHelper.kt:
```diff
@@ -0,0 +1,36 @@
+import android.content.Context
+import androidx.appcompat.app.AppCompatDelegate
+
+object ThemeHelper {
```
Логику работы с темой приложения лучше не хранить в статическом классе с неймингом Helper (гугл не рекомендует этот нейминг к использованию) - тут лучше тоже вынести эту логику в интерфейс интерактора и репозитория

ArturNurtdinov replied:
В след спринте также будет требование, что работа с SharedPrefs должна происходить только в data слое - по сути код с хранением темы приложения должен находиться в соответствующем репозитории

ArturNurtdinov reviewed (commented):
Отступ длительности трека не соответствует макетам в списке треков:

<img width="852" height="283" alt="image" src="https://github.com/user-attachments/assets/206694fa-20eb-4a3b-ab67-ed5621f1b619" />

Макет:

<img width="1169" height="454" alt="image" src="https://github.com/user-attachments/assets/7a65c2be-8931-460b-8ae1-74d4cae8be8f" />

Ulyanov-VN committed:
Домашнее задание Модуль №15 (устранение замечаний)
1.  Приложение работает корректно: возможности, предоставляемые приложением до рефакторинга, предоставляются в том же объёме и качестве - Отступ длительности трека не соответствует макетам в списке треков

Замечание устранено, приведено в соответствие с макетом

2.  Доступ к данным должен осуществляться через репозитории, которые инкапсулируют логику получения и сохранения данных - Доступ к значению темы приложения в SharedPreferences осуществляется через класс ThemeHelper, нужно реализовать репозиторий для хранения данных

Замечания устранил и отработал комментарии выданные в гите

3. Работа с интеракторами организована через интерфейсы - Отсутствуют интерфейсы для FormatTimeInteractor, GetCountryNameInteractor, GetCoverArtworkInteractor, GetReleaseYearInteractor,

Добавлены интерфейсы

ArturNurtdinov commented on app/src/main/java/com/example/playlistmaker/presentation/activities/BaseActivity.kt:
```diff
@@ -2,15 +2,25 @@ package com.example.playlistmaker.presentation.activities
+import com.example.playlistmaker.domain.interactor.ManageThemeInteractor
 import com.google.android.material.bottomnavigation.BottomNavigationView
 
 abstract class BaseActivity : AppCompatActivity() {
```
Абстрактные бейз классы лучше не использовать, т.к. такие классы обычно быстро разрастаются и их становится тяжело наследовать. Лучше все-таки применение темы перенести в класс-наследник Application, а от BaseActivity - избавиться

К тому же такие Base классы нарушают принцип Single Responsibility - даже этот класс одновременно и панель навигации настраивает и применяет тему

ArturNurtdinov commented on app/src/main/java/com/example/playlistmaker/presentation/activities/BaseActivity.kt:
```diff
@@ -23,6 +33,28 @@ abstract class BaseActivity : AppCompatActivity() {
+        return newState
+    }
+
     private fun setupBottomNavigation(bnv: BottomNavigationView) {
```
Панель навигации будет настраиваться не в каждой активити, а только в одной рут активити
